File: web/public/js/hall/hall.js
/*
 * @author sheppard(ysf1026@gmail.com) 2014-01-02
 */

var socket;
var msgs = [];

$(document).ready(function() {
  connect();

  var btn_test = $('#btn-test');
  btn_test.click(function() {
    var content = btn_test.text(); 
    socket.send(content);
    return false;
  });
});

function connect() {
  var userName = $('#userName').text();
  socket = new WebSocket('ws://'+window.location.host+'/hall/hall/socket?user=' + userName);

  socket.onmessage = function(event) {
    var msg = JSON.parse(event.data);
    addToPanel(msg.Name, msg.Content);
    refreshPanel();
  };

  socket.onclose = function() {
    console.log('disconnected');
  };
};

function addToPanel(prefix, msg) {
  msgs.push({prefix: prefix, msg: msg});
};

function refreshPanel() {
  var panel = $('#panel-chat');
  panel.empty();
  _.each(msgs, function(item) {
    panel.append('<strong>' + item.prefix + ':</strong> ');
    panel.append(item.msg).append('<br>');
  });
};

function refreshRoomList() {
  var rooms = $('#list-room');
}

File: app/logic/hall/hall.go
// Author: sheppard(ysf1026@gmail.com) 2014-01-12

package hall

import (
	"fmt"
)

type User interface {
	Send(string)
}

var (
	users map[string] User

	broadcast chan string
)

func init() {
	users = make(map[string] User)
	broadcast = make(chan string)

	go Tick()
}

func AddUser(uid string, u User) bool {
	if _, ok := users[uid]; ok {
		return false
	}
	users[uid] = u
	return true
}

func DelUser(uid string) {
	delete(users, uid)
}

func Broadcast(msg string) {
	broadcast <- msg
}

func Tick() {
	for {
		msg := <-broadcast
		fmt.Println("Hall broadcast msg:", msg)
		for _, u := range users {
			u.Send(msg)
		}
	}
}

